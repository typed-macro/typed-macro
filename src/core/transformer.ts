import { CallExpression, File } from '@babel/types'
import traverse, { NodePath } from '@babel/traverse'
import { parse, ParserPlugin } from '@babel/parser'
import generate from '@babel/generator'
import { nodeLoc } from '@/common'
import { NamespacedMacros } from './exports'
import { Macro } from './macro'
import { CURRENT_MACRO_CALL_VERSION, versioned } from '@/core/version'
import {
  findImportedMacros,
  findProgramPath,
  getCalledMacro,
} from '@/core/helper/traverse'
import { createState, State } from './helper/state'
import { isError, isPromise } from '@/common'

export type TransformerOptions = {
  /**
   * The max recursion for applying macros, default to 5.
   *
   * After reached the maxRecursions, plugin will throw out an error.
   *
   * It's usually caused by forgetting to remove/modify the call expression
   * in macros.
   */
  maxRecursions?: number
  /**
   * Babel plugins to be applied during parsing.
   *
   * 'typescript', 'jsx' and 'importMeta'
   * are included by default and cannot be removed.
   *
   * Some plugins are enabled automatically by the latest babel parser
   * because they have been regarded as part of the language.
   *
   * @see https://babeljs.io/docs/en/babel-parser#plugins
   * @see https://babeljs.io/docs/en/babel-parser#latest-ecmascript-features
   */
  parserPlugins?: ParserPlugin[]
}

export type TransformerContext = {
  code: string
  filepath: string
  ssr?: boolean
  dev: boolean
}

export type Transformer = (
  ctx: TransformerContext,
  macros: NamespacedMacros
) => string | undefined

export function createTransformer({
  parserPlugins = [],
  maxRecursions = 0,
}: TransformerOptions): Transformer {
  maxRecursions = maxRecursions > 0 ? maxRecursions : 5
  parserPlugins = ['typescript', 'jsx', 'importMeta', ...parserPlugins]

  return ({ code, filepath, ssr = false, dev }, macros) => {
    const ast = parse(code, {
      sourceType: 'module',
      plugins: parserPlugins,
    })

    // keep import statements in dev mode so can invalidate macro modules
    const collect = () => findImportedMacros(ast, macros, dev)

    const importedMacros = collect()
    if (!importedMacros.length) return

    const transformState = createState()

    let loopCount = 0
    while (
      applyMacros({
        code,
        filepath,
        ast,
        ssr,
        importedMacros,
        transformState,
      }) &&
      importedMacros.push(...collect())
    ) {
      if (loopCount++ >= maxRecursions)
        throw new Error(
          `Reached the maximum recursion when apply macros on '${filepath}'`
        )
    }

    return generate(ast, {
      retainLines: true,
    }).code
  }
}

type ImportedAsNS = {
  importedAsNamespace: true
  namespace: string
  local: string
  macros: Macro[]
}

type ImportedAsNamed = {
  importedAsNamespace: false
  namespace: string
  local: string
  // not always macro
  macro?: Macro
}

type ImportedMacro = ImportedAsNS | ImportedAsNamed

type ApplyContext = {
  code: string
  filepath: string
  ast: File
  ssr: boolean

  importedMacros: ImportedMacro[]
  transformState: State
}

export function applyMacros({
  code,
  filepath,
  ast,
  importedMacros,
  ssr,
  transformState,
}: ApplyContext): boolean {
  let applied = false
  const program = findProgramPath(ast)
  const traversalState = createState()

  const process = (path: NodePath<CallExpression>) => {
    const macroToApply = getCalledMacro(path.get('callee'), importedMacros)
    if (!macroToApply) return
    if (typeof macroToApply === 'string')
      throw new Error(
        `Macro '${macroToApply}' is not existed but is called in '${filepath}'`
      )
    try {
      macroToApply(
        versioned(
          {
            code,
            filepath,
            path,
            ssr,
            ast,
            transformState,
            traversalState,
            importedMacros,
            program,
          },
          CURRENT_MACRO_CALL_VERSION
        )
      )
    } catch (e: unknown) {
      if (isError(e)) {
        // throw errors
        throw new Error(
          `Error when apply macro '${
            macroToApply.name
          }' in '${filepath}' near ${nodeLoc(path.node)}:\n ${e}`
        )
      } else if (isPromise(e)) {
        // expand nested macros first
        path.traverse({
          CallExpression(p) {
            process(p)
          },
        })
        process(path)
      }
    } finally {
      applied = true
    }
  }

  traverse(ast, {
    CallExpression(path) {
      process(path)
    },
  })

  return applied
}
