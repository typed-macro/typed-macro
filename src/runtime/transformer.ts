import {
  CallExpression,
  File,
  ImportDeclaration,
  isIdentifier,
  isImportDefaultSpecifier,
  isImportNamespaceSpecifier,
  isMemberExpression,
  isStringLiteral,
  Node,
} from '@babel/types'
import traverse, { NodePath } from '@babel/traverse'
import template from '@babel/template'
import { parse, ParserPlugin } from '@babel/parser'
import generate from '@babel/generator'
import { nodeLoc } from '@/common'
import { Macro, MacroHelper, NamespacedMacros } from './types'
import { getTransformHelper } from '@/helper/transform'
import { getPathHelper } from '@/helper/path'
import { getStateHelper, State } from '@/helper/state'
import { BABEL_TOOLS } from '@/helper/babel'

export type TransformerOptions = {
  maxRecursion?: number
  parserPlugins?: ParserPlugin[]
}

export type TransformerContext = {
  code: string
  id: string
  ssr?: boolean
  dev: boolean
}

export type Transformer = (
  ctx: TransformerContext,
  macros: NamespacedMacros
) => string | undefined

export function getTransformer({
  parserPlugins = [],
  maxRecursion = 0,
}: TransformerOptions): Transformer {
  maxRecursion = maxRecursion > 0 ? maxRecursion : 5
  parserPlugins = ['typescript', 'jsx', ...parserPlugins]

  return ({ code, id, ssr = false, dev }, macros) => {
    const ast = parse(code, {
      sourceType: 'module',
      plugins: parserPlugins,
    })

    const collect = () =>
      collectImportedMacros(
        ast,
        macros,
        // Note: keep import statements in dev mode so can invalidate macro modules
        dev /* keep import stmt */
      )
    const importedMacros = collect()
    if (!importedMacros.length) return

    const ctx: ApplyContext = {
      code,
      filepath: id,
      ast,
      ssr,
      importedMacros,
      state: {
        thisTurn: {},
        crossTurn: {},
      },
    }

    let loopCount = 0
    while (loopCount++ < maxRecursion && applyMacros(ctx)) {
      if (loopCount === maxRecursion)
        throw new Error(
          `Reached the maximum recursion, please check macros applied in file ${id}`
        )
      ctx.importedMacros.push(...collect())
      // clear this turn's state
      ctx.state.thisTurn = {}
    }

    return generate(ast, {
      retainLines: true,
    }).code
  }
}

type ImportedAsNS = {
  importedAsNamespace: true
  namespace: string
  local: string
  macros: Macro[]
}

type ImportedAsNamed = {
  importedAsNamespace: false
  namespace: string
  local: string
  // not always macro
  macro?: Macro
}

type ImportedMacro = ImportedAsNS | ImportedAsNamed

export function collectImportedMacros(
  ast: Node,
  macros: NamespacedMacros,
  keepImportStmt = false
) {
  const importedMacros: ImportedMacro[] = []
  const namespaces = Object.keys(macros)
  {
    const importPaths: NodePath<ImportDeclaration>[] = []
    traverse(ast, {
      ImportDeclaration(path) {
        const ns = path.node.source.value
        if (!namespaces.includes(ns)) return

        // case - import 'a'
        if (!path.node.specifiers.length) return

        path.node.specifiers.forEach((s) => {
          if (isImportDefaultSpecifier(s) || isImportNamespaceSpecifier(s)) {
            // case - import a from 'a'
            // or
            // case - import * as a from 'a'
            importedMacros.push({
              importedAsNamespace: true,
              namespace: path.node.source.value,
              local: s.local.name,
              macros: macros[ns],
            })
          } else {
            // case - import { a as a } from 'a'
            const exportName = isIdentifier(s.imported)
              ? s.imported.name
              : s.imported.value

            importedMacros.push({
              importedAsNamespace: false,
              namespace: path.node.source.value,
              local: s.local.name,
              macro: macros[ns].find((m) => m.name === exportName),
            })
          }
        })
        importPaths.push(path)
      },
    })
    if (keepImportStmt) {
      importPaths.forEach((p) =>
        p.replaceWith(template.statement.ast(`import '${p.node.source.value}'`))
      )
    } else {
      importPaths.forEach((p) => p.remove())
    }
  }
  return importedMacros
}

// find macros to be applied from call expr, returns a
// - Macro: the macro to be applied
// - string: is a macro call but no macro found
// - undefined: not a macro call
export function findCalledMacro(
  callee: CallExpression['callee'],
  importedMacros: Array<ImportedMacro>
) {
  if (isMemberExpression(callee)) {
    const ns = callee.object
    if (isIdentifier(ns)) {
      const maybeMacro = importedMacros.find(
        (m) => m.importedAsNamespace && m.local === ns.name
      ) as ImportedAsNS
      if (!maybeMacro) return
      const method = callee.property
      if (isIdentifier(method)) {
        // case - namespace.method()
        return (
          maybeMacro.macros.find((m) => m.name === method.name) || method.name
        )
      } else if (isStringLiteral(method)) {
        // case - namespace['method']()
        return (
          maybeMacro.macros.find((m) => m.name === method.value) || method.value
        )
      }
    }
  } else if (isIdentifier(callee)) {
    // case - method()
    const maybeMacro = importedMacros.find(
      (m) => !m.importedAsNamespace && m.local === callee.name
    ) as ImportedAsNamed
    if (!maybeMacro) return
    return maybeMacro.macro || callee.name
  }
}

type ApplyContext = {
  code: string
  filepath: string
  ast: File
  ssr: boolean

  importedMacros: ImportedMacro[]
  state: State
}

export function applyMacros({
  code,
  filepath,
  ast,
  importedMacros,
  ssr,
  state,
}: ApplyContext): boolean {
  let applied = false

  const helper: MacroHelper = Object.freeze({
    ...getTransformHelper(ast),
    ...getPathHelper(filepath),
    ...getStateHelper(state),
  })

  traverse(ast, {
    CallExpression(path) {
      const macroToApply = findCalledMacro(path.node.callee, importedMacros)
      if (!macroToApply) return

      if (typeof macroToApply === 'string')
        throw new Error(
          `Macro ${macroToApply} is not existed but is called in ${filepath}`
        )

      try {
        macroToApply.apply(
          {
            code,
            filepath,
            path,
            ssr,
            args: path.node.arguments,
          },
          BABEL_TOOLS,
          helper
        )
      } catch (e) {
        throw new Error(
          `Error when apply macro ${
            macroToApply.name
          } in ${filepath} near ${nodeLoc(path.node)}:\n ${e}`
        )
      } finally {
        applied = true
      }
    },
  })

  return applied
}
