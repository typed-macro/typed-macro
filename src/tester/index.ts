export * from './transformer'
export * from './typeRenderer'
