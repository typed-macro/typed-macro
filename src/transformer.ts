import {
  CallExpression,
  File,
  ImportDeclaration,
  isIdentifier,
  isImportDefaultSpecifier,
  isImportNamespaceSpecifier,
  isMemberExpression,
  isStringLiteral,
  Node,
} from '@babel/types'
import traverse, { NodePath } from '@babel/traverse'
import template from '@babel/template'
import { Macro } from '@/macro'
import { BABEL_TOOLS, getHelper } from '@/helper'
import { parse, ParserPlugin } from '@babel/parser'
import generate from '@babel/generator'
import { nodeLoc } from '@/common'

export type NamespacedMacros = { [namespace: string]: Macro[] }

export type TransformerOptions = {
  macros: NamespacedMacros

  maxRecursion: number
  parserPlugins: ParserPlugin[]
}

export function getTransformer({
  parserPlugins,
  maxRecursion,
  macros,
}: TransformerOptions) {
  throwErrorIfConflict(macros)

  return (code: string, id: string, dev: boolean) => {
    const ast = parse(code, {
      sourceType: 'module',
      plugins: parserPlugins,
    })

    const collect = () =>
      collectImportedMacros(
        ast,
        macros,
        // Note: keep import statements in dev mode so can invalidate macro modules
        dev /* keep import stmt */
      )
    const importedMacros = collect()
    if (!importedMacros.length) return

    const applyCtx: ApplyContext = {
      code,
      filepath: id,
      ast,
      importedMacros,
    }

    let loopCount = 0
    while (loopCount < maxRecursion) {
      const { applied, recollectMacros } = applyMacros(applyCtx)
      if (!applied) break
      if (recollectMacros) importedMacros.push(...collect())

      loopCount++
      if (loopCount === maxRecursion)
        throw new Error(
          `Reached the maximum recursion, please check macros applied in file ${id}`
        )
    }

    return generate(ast, {
      retainLines: true,
    }).code
  }
}

export function throwErrorIfConflict(macros: NamespacedMacros) {
  Object.keys(macros).forEach((ns) => {
    const mem = Object.create(null)
    macros[ns].forEach((m) => {
      if (mem[m.name]) {
        throw new Error(
          `Error when loading macros: a macro with name '${m.name}' in '${ns}' already existed`
        )
      }
      mem[m.name] = 1
    })
  })
}

type ImportedAsNS = {
  importedAsNamespace: true
  namespace: string
  local: string
  macros: Macro[]
}

type ImportedAsFn = {
  importedAsNamespace: false
  namespace: string
  local: string
  macro: Macro
}

type ImportedMacro = ImportedAsNS | ImportedAsFn

export function collectImportedMacros(
  ast: Node,
  macros: NamespacedMacros,
  keepImportStmt = false
) {
  const importedMacros: ImportedMacro[] = []
  const namespaces = Object.keys(macros)
  {
    const importPaths: NodePath<ImportDeclaration>[] = []
    traverse(ast, {
      ImportDeclaration(path) {
        const ns = path.node.source.value
        if (!namespaces.includes(ns)) return

        // case - import 'a'
        if (!path.node.specifiers.length) return

        path.node.specifiers.forEach((s) => {
          if (isImportDefaultSpecifier(s) || isImportNamespaceSpecifier(s)) {
            // case - import a from 'a'
            // or
            // case - import * as a from 'a'
            importedMacros.push({
              importedAsNamespace: true,
              namespace: path.node.source.value,
              local: s.local.name,
              macros: macros[ns],
            })
          } else {
            // case - import { a as a } from 'a'
            const exportName = isIdentifier(s.imported)
              ? s.imported.name
              : s.imported.value
            const macro = macros[ns].find((m) => m.name === exportName)
            if (!macro)
              throw new Error(
                `Macro ${exportName} is not existed but is imported from '${ns}'`
              )
            importedMacros.push({
              importedAsNamespace: false,
              namespace: path.node.source.value,
              local: s.local.name,
              macro,
            })
          }
        })
        importPaths.push(path)
      },
    })
    if (keepImportStmt) {
      importPaths.forEach((p) =>
        p.replaceWith(template.statement.ast(`import '${p.node.source.value}'`))
      )
    } else {
      importPaths.forEach((p) => p.remove())
    }
  }
  return importedMacros
}

// find macros to be applied from call expr, returns a
// - Macro: has origin
// - string: has origin name but no origin macro found
// - undefined: no origin name found
export function findCalledMacro(
  callee: CallExpression['callee'],
  importedMacros: Array<ImportedMacro>
) {
  if (isMemberExpression(callee)) {
    const ns = callee.object
    if (isIdentifier(ns)) {
      const maybeMacro = importedMacros.find(
        (m) => m.importedAsNamespace && m.local === ns.name
      ) as ImportedAsNS
      if (!maybeMacro) return
      const method = callee.property
      if (isIdentifier(method)) {
        // case - namespace.method()
        return (
          maybeMacro.macros.find((m) => m.name === method.name) || method.name
        )
      } else if (isStringLiteral(method)) {
        // case - namespace['method']()
        return (
          maybeMacro.macros.find((m) => m.name === method.value) || method.value
        )
      }
    }
  } else if (isIdentifier(callee)) {
    // case - method()
    const maybeMacro = importedMacros.find(
      (m) => !m.importedAsNamespace && m.local === callee.name
    ) as ImportedAsFn
    if (!maybeMacro) return
    return maybeMacro.macro
  }
}

type ApplyContext = {
  code: string
  filepath: string
  ast: File

  importedMacros: ImportedMacro[]
}

type ApplyResult = {
  applied: boolean
  recollectMacros: boolean
}

export function applyMacros({
  code,
  filepath,
  ast,
  importedMacros,
}: ApplyContext): ApplyResult {
  const result: ApplyResult = {
    applied: false,
    recollectMacros: false,
  }

  const helper = Object.freeze({
    forceRecollectMacros: () => (result.recollectMacros = true),
    ...getHelper(filepath, ast),
  })

  traverse(ast, {
    CallExpression(path) {
      const macroToApply = findCalledMacro(path.node.callee, importedMacros)
      if (!macroToApply) return

      if (typeof macroToApply === 'string')
        throw new Error(
          `Macro ${macroToApply} is not existed but is called in ${filepath}`
        )

      try {
        macroToApply.apply(
          {
            code,
            filepath,
            path,
            args: path.node.arguments,
          },
          BABEL_TOOLS,
          helper
        )
      } catch (e) {
        throw new Error(
          `Error when apply macro ${
            macroToApply.name
          } in ${filepath} near ${nodeLoc(path.node)}:\n ${e}`
        )
      } finally {
        result.applied = true
      }
    },
  })

  return result
}
