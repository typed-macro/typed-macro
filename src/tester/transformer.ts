import { NamespacedMacros } from '@/core/exports'
import { createTransformer, TransformerOptions } from '@/core/transformer'

export type TestTransformerContext = {
  /**
   * Source code.
   */
  code: string

  /**
   * Path of the source file.
   *
   * Default to 'test.tsx'
   */
  filepath?: string

  /**
   * Is in SSR mode.
   *
   * Default to false.
   */
  ssr?: boolean

  /**
   * Is in dev mode.
   *
   * Default to false.
   *
   * If in dev mode, the import statements that imports macros
   * will be retained and rewritten by the runtime
   * to track the module's dependence on the macro.
   */
  dev?: boolean
}

export type TestTransformer = {
  (ctx: TestTransformerContext, macros: NamespacedMacros): string
  (code: string, macros: NamespacedMacros): string
}

/**
 * Create a transformer to help test macros.
 *
 * e.g.
 * ```typescript
 * const aMacro = defineMacro(...)
 *   .withSignature(...)
 *   .withHandler(...)
 *
 * const transform = createTestTransformer()
 *
 * expect(
 *   transform(`some code...`, { '@ns': [aMacro] })
 * ).toBe(`some code...`)
 * ```
 *
 * @param options the same as the one of transformer
 */
export function createTestTransformer(
  options: TransformerOptions = {}
): TestTransformer {
  const transform = createTransformer(options)
  return (input, macros) => {
    const {
      code,
      filepath = 'test.tsx',
      ssr = false,
      dev = false,
    } = typeof input === 'string' ? { code: input } : input
    return transform({ code, filepath, ssr, dev }, macros) || code
  }
}
