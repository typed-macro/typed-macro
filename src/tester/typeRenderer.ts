import { NamespacedExportable, normalizeExports } from '@/core/exports'
import { renderTypes } from '@/core/typeRenderer'

export type TestTypeRenderer = (exports: NamespacedExportable) => string

/**
 * Create a type renderer to help test the type of macros/modules.
 *
 * e.g.
 * ```typescript
 * const aMacro = defineMacro(...)
 *   .withSignature(...)
 *   .withHandler(...)
 *
 * const renderType = createTestTypeRenderer()
 *
 * expect(
 *   renderType({
 *     '@ns': {
 *       macros: [aMacro]
 *       customTypes: `some types...`
 *     },
 *     '@another': {
 *       code: `some code...`
 *       customTypes: `some types...`
 *     }
 *   })
 * ).toMatchSnapshot()
 * ```
 */
export function createTestTypeRenderer(): TestTypeRenderer {
  return (exports) => {
    const normalized = normalizeExports(exports)
    return renderTypes(normalized.types)
  }
}
