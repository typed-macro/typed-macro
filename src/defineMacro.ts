import { validateFnName } from '@/common'
import { Macro, MacroHandler, MacroWithType } from '@/runtime/types'

type MacroBuilder = {
  /**
   * Add custom type definition, it will be written to .d.ts before macro signature.
   */
  withCustomType: (typeDefinition: string) => Omit<MacroBuilder, 'withHandler'>
  /**
   * Add signature of macro. one macro requires at least one signature.
   * @param signature function signature, like '(s: string): void'
   * @param comment comment for the signature, will be written to .d.ts
   * before the corresponding function signature.
   */
  withSignature: (
    signature: string,
    comment?: string
  ) => Omit<MacroBuilder, 'withCustomType'>
  /**
   * Set the transform handler and get the macro.
   */
  withHandler: (handler: MacroHandler) => Macro
}

type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  customTypes: string[]
}

/**
 * Define a macro.
 * @param name the name of the macro, should be a valid identifier name.
 * @return macro builder.
 */
export function defineMacro(name: string): Omit<MacroBuilder, 'withHandler'> {
  if (!validateFnName(name))
    throw new Error(`'${name}' is not a valid macro name!`)

  const meta: MacroMeta = {
    signatures: [],
    customTypes: [],
  }

  const builder: MacroBuilder = {
    withCustomType(typeDefinition) {
      meta.customTypes.push(typeDefinition)
      return builder
    },
    withSignature(signature, comment) {
      meta.signatures.push({ signature, comment })
      return builder
    },
    withHandler(handler) {
      if (meta.signatures.length === 0)
        throw new Error(
          `Please call .withSignature() before .withHandler() to specify at least one signature for macro '${name}'`
        )
      return {
        name,
        apply: handler,
        __types: renderMacroType(name, meta),
      } as MacroWithType
    },
  }

  return builder
}

function renderMacroType(name: string, meta: MacroMeta) {
  return [
    meta.customTypes.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
