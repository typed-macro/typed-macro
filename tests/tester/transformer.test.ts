import { createTestTransformer } from '@/tester/transformer'
import { defineMacro } from '@/defineMacro'

describe('createTestTransformer()', () => {
  const transform = createTestTransformer()

  it('should return original code if no transformation', () => {
    // pass code string
    expect(transform(`console.log('a')`, {})).toBe(`console.log('a')`)
    // pass ctx
    expect(transform({ code: `console.log('a')` }, {})).toBe(`console.log('a')`)
  })

  it('should work', () => {
    const echoMacro = defineMacro('echo')
      .withSignature('')
      .withHandler(({ path }, { template }) => {
        path.replaceWith(template.expression.ast(`console.log('a')`))
      })

    const macros = {
      '@macro': [echoMacro],
    }

    expect(
      transform(`import { echo } from '@macro'; echo()`, macros)
    ).toMatchSnapshot()
    expect(
      transform(
        { code: `import { echo } from '@macro'; echo()`, dev: true },
        macros
      )
    ).toMatchSnapshot()
  })
})
