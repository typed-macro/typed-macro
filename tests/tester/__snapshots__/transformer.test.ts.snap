// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTestTransformer() should work 1`] = `"console.log('a');"`;

exports[`createTestTransformer() should work 2`] = `"import '@macro';console.log('a');"`;
